rules/okta_rules/okta_potentially_stolen_session.py

```diff
@@ -0,0 +1,83 @@
+from ast import literal_eval
+from datetime import datetime, timedelta
+from difflib import SequenceMatcher
+
+from panther_base_helpers import deep_get, okta_alert_context
+from panther_oss_helpers import get_string_set, put_string_set, set_key_expiration
+
+PREVIOUS_SESSION = {}
+FUZZ_RATIO_MIN = 0.95
+
+
+def rule(event):
+    # pylint: disable=global-statement
+    # ensure previous session info is avaialable in the alert_context for investigation
+    global PREVIOUS_SESSION
+
+    # Filter only on app access and session start events
+    if (
+        event.get("eventType") == "user.authentication.sso"  # user opened app
+        or event.get("eventType") == "user.session.start"  # user loged in
+    ):
+
+        # lookup if we've previously stored the session cookie
+        PREVIOUS_SESSION = get_string_set(
+            deep_get(event, "authenticationContext", "externalSessionId"),
+        )
+
+        # For unit test mocks we need to eval the string to a set
+        if isinstance(PREVIOUS_SESSION, str):
+            PREVIOUS_SESSION = literal_eval(PREVIOUS_SESSION)
+
+        # if the session cookie was seen before
+        if PREVIOUS_SESSION is not None:
+            # we use a fuzz match to compare the current and prev user agent.
+            # We cannot do a direct match since Okta can occasionally maintain
+            # a session across browser upgrades.
+
+            # the user-agent is the longest string in the set
+            prev_ua = [x for x in PREVIOUS_SESSION if len(x) > 25]
```

Comment:
I don't believe this is going to work as intended once there's `> 1` data to compare, or if there are malicious UAs coming in. I think that `panther_oss_helpers.*counter` methods against a single specific field in the event will lead to more deterministic behavior.